lib/wasmex/components.ex

```diff
@@ -0,0 +1,39 @@
+defmodule Wasmex.Components do
```

Comment:
interesting, do you plan to have a different GenServer than our default Wasmex one? What's the reasoning?
Do you plan to eventually go full-component and remove the default one? Do you think that over time we need more specialised GenServers? Do you think they are implemented so different that they do not have much overlap?

I'm open to ideas, but want to understand the reasoning :) My first instinct is that this and the old GenServer will probably look very much the same and end up implementing almost the same thing. Maybe they should be the same? But then, if every handle_call is a big switch, we might end up very spaghetti and it's better to separate both things (plain wasm, components) from the start 🤔  i'm very interested in your thoughts

Reply:
uhhh, while we talk components. There was a very helpful comment by José a while back where he suggested our GenServer should queue incoming requests in case the client is busy running some WASM.

https://github.com/tessi/wasmex/issues/391#issuecomment-1374279165

maybe we should implement that idea from the start in this new GenServer (but maybe in a separate PR, if you prefer)

Reply:
I initially went this direction based on what I saw in the wasmtime codebase, where they have separate packages for components and non-component wasm. I also need a different version of store, and there is a different version WASI. It might be possible to merge into a single GenServer module, but the more I look at the code I'm less confident of trying do this. Keeping them separate also lets me get something out while I'm gradually adding features (linking, etc) that you already have in the main `Wasmex` genserver.

Long term, I like the idea of allowing the user to define component specific genservers and building macros to give them more specific, high-level functions from the wit definitions. Eg, a `TodoList` genserver module would have an `add_todo` fuction which took a pid and the arguments to pass to the exported function, with the correct arity rather than having to bundle the args into a list.

For the queuing idea I kinda of want to wait on that. It strikes me as something we or someone else could layer in if they need it maybe?

Reply:
👍 on going separate
👍 on queueing later
and let's talk about the macro idea separate (my belly feeling points towards code generator and not towards macros, but i didn't explore this yet. so that's a very weak opinion)

Reply:
I don't have a strong opinion here, code generators vs macros. Might be fun to check out some of the work Ash has been doing with their Igniter thing. It might make sense for this to be a separate package even